src/pages/signin/ValidateCodeForm.js

```diff
@@ -0,0 +1,230 @@
+        this.clearSignInData = this.clearSignInData.bind(this);
+
+        this.state = {
+            formError: false,
+            validateCode: '',
+            twoFactorAuthCode: '',
+        };
+    }
+
+    componentDidMount() {
+        if (!canFocusInputOnScreenFocus() || !this.inputValidateCode || !this.props.isVisible) {
+            return;
+        }
+        this.inputValidateCode.focus();
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if (!prevProps.isVisible && this.props.isVisible) {
+            this.inputValidateCode.focus();
+        }
+        if (prevProps.isVisible && !this.props.isVisible && this.state.validateCode) {
+            this.clearValidateCode();
+        }
+        if (!prevProps.account.requiresTwoFactorAuth && this.props.account.requiresTwoFactorAuth) {
+            this.input2FA.focus();
+        }
+        if (prevState.twoFactorAuthCode !== this.state.twoFactorAuthCode && this.state.twoFactorAuthCode.length === CONST.TFA_CODE_LENGTH) {
+            this.validateAndSubmitForm();
+        }
+    }
+
+    /**
+     * Clear Validate Code from the state
+     */
+    clearValidateCode() {
+        this.setState({validateCode: ''}, this.inputValidateCode.clear);
+    }
+
+    /**
+     * Trigger the reset validate code flow and ensure the 2FA input field is reset to avoid it being permanently hidden
+     */
+    resetValidateCode() {
+        if (this.input2FA) {
+            this.setState({twoFactorAuthCode: ''}, this.input2FA.clear);
+        }
+        this.setState({formError: false});
+        User.resendValidateCode(this.props.credentials.login, true);
+    }
+
+    /**
+    * Clears local and Onyx sign in states
+    */
+    clearSignInData() {
+        this.setState({twoFactorAuthCode: '', formError: false});
+        Session.clearSignInData();
+    }
+
+    /**
+     * Check that all the form fields are valid, then trigger the submit callback
+     */
+    validateAndSubmitForm() {
+        if (!this.state.validateCode.trim()) {
+            this.setState({formError: 'validateCodeForm.error.pleaseFillMagicCode'});
+            return;
+        }
+
+        if (!ValidationUtils.isValidValidateCode(this.state.validateCode)) {
+            this.setState({formError: 'validateCodeForm.error.incorrectMagicCode'});
+            return;
+        }
+
+        if (this.props.account.requiresTwoFactorAuth && !this.state.twoFactorAuthCode.trim()) {
+            this.setState({formError: 'validateCodeForm.error.pleaseFillTwoFactorAuth'});
+            return;
+        }
+
+        this.setState({
+            formError: null,
+        });
+
+        Session.signIn('', this.state.validateCode, this.state.twoFactorAuthCode);
+    }
+
+    render() {
+        return (
+            <>
+                {/* At this point, if we know the account requires 2FA we already successfully authenticated */}
+                {this.props.account.requiresTwoFactorAuth ? (
```

Comment:
FYI I wanted to refactor this whole thing to use the Form component, but it got a bit iffy with clearing the 2FA input & auto-focusing some fields, so I decided this would be better as a follow-up so we can get this out the door